text/0340-firehose-l2.md

```diff
@@ -0,0 +1,944 @@
+- `grantWriteData(principal)` - grants the principal write access to the data plane
+- `grantFullAccess(principal)` - grants principal full access to the delivery stream
+
+### Control Plane Read Permissions
+
+Grant `read` access to the control plane of a delivery stream by calling the `grantRead()` method.
+
+```ts fixture=with-delivery-stream
+import * as iam from '@aws-cdk/aws-iam';
+const lambdaRole = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+}
+
+// Give the role permissions to read information about the delivery stream
+deliveryStream.grantRead(lambdaRole);
+```
+
+The following read permissions are provided to a service principal by the `grantRead()` method:
+
+- `firehose:DescribeDeliveryStream`
+- `firehose:ListDeliveryStreams`
+- `firehose:ListTagsForDeliveryStream`
+
+### Control Plane Write Permissions
+
+Grant `write` access to the control plane of a delivery stream by calling the `grantWrite()` method.
+
+```ts fixture=with-delivery-stream
+import * as iam from '@aws-cdk/aws-iam';
+const lambdaRole = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+}
+
+// Give the role permissions to modify the delivery stream
+deliveryStream.grantWrite(lambdaRole);
+```
+
+The following write permissions are provided to a service principal by the `grantWrite()` method:
+
+- `firehose:DeleteDeliveryStream`
+- `firehose:StartDeliveryStreamEncryption`
+- `firehose:StopDeliveryStreamEncryption`
+- `firehose:UpdateDestination`
+
+### Data Plane Write Permissions
+
+Grant `write` access to the data plane of a delivery stream by calling the `grantWriteData()` method.
+
+```ts fixture=with-delivery-stream
+import * as iam from '@aws-cdk/aws-iam';
+const lambdaRole = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+}
+
+// Give the role permissions to write data to the delivery stream
+deliveryStream.grantWriteData(lambdaRole);
+```
+
+The following write permissions are provided to a service principal by the `grantWriteData()` method:
+
+- `firehose:PutRecord`
+- `firehose:PutRecordBatch`
+
+### Custom Permissions
+
+You can add any set of permissions to a delivery stream by calling the `grant()` method.
+
+```ts fixture=with-delivery-stream
+import * as iam from '@aws-cdk/aws-iam';
+const user = new iam.User(this, 'User');
+
+// give user permissions to update destination
+deliveryStream.grant(user, 'firehose:UpdateDestination');
+```
+
+---
+
+# Amazon Kinesis Data Firehose Destinations Library
+
+This library provides constructs for adding destinations to a Amazon Kinesis Data Firehose
+delivery stream.  Destinations can be added by specifying the `destination` prop when
+creating a delivery stream.
+
+See Amazon Kinesis Data Firehose module README for usage examples.
+
+If further customization is required, use `HttpDestination` from this package or implement
+`firehose.IDestination`.
```

Comment:
If you really mean this, be prepared to put `@subclassable` on `IDestination`, which means you never get to change the interface ever again :).

Reply:
Could we still change the interface while the module is experimental? And is it uncommon for customers to create their own implementations of CDK interfaces?

Reply:
I think this sentence was a bit of a hedge – if `HttpDestination` is built correctly, then it should support *any* custom HTTP endpoint and any further customization/defaulting by customers should subclass `HttpDestination`